packages/background/src/utils/swaps/1inch.ts

```diff
@@ -106,13 +119,129 @@ export interface OneInchSwapRequestResponse {
     toToken: BasicToken;
     fromTokenAmount: string; // Input amount of fromToken in minimal divisible units
     toTokenAmount: string; // Result amount of toToken in minimal divisible units
-    protocols: unknown[]; // Route of the trade
-    tx: {
-        from: string; // Transactions will be sent from this address
-        to: string; // Transactions will be sent to our contract address
-        data: string; // Bytes of call data
-        value: string; // Amount of ETH (in wei) will be sent to the contract address
-        gas: number; // Estimated amount of the gas limit, increase this value by 25%
-        gasPrice: string; // Gas price in wei
-    };
+    protocols?: unknown[]; // Route of the trade
+    tx: SwapTxMeta;
+}
+
+export interface SwapTxMeta {
+    from: string; // Transactions will be sent from this address
+    to: string; // Transactions will be sent to our contract address
+    data: string; // Bytes of call data
+    value: string; // Amount of ETH (in wei) will be sent to the contract address
+    gas: number; // Estimated amount of the gas limit, increase this value by 25%
+    gasPrice: string; // Gas price in wei
 }
+
+export const OneInchService = {
+    parseQuoteParams({
+        fromToken,
+        toToken,
+        amount,
+    }: SwapQuoteParams): OneInchSwapQuoteParams {
+        return {
+            fromTokenAddress:
+                fromToken.address === '0x0'
+                    ? SWAP_NATIVE_ADDRESS
+                    : fromToken.address,
+            toTokenAddress:
+                toToken.address === '0x0'
+                    ? SWAP_NATIVE_ADDRESS
+                    : toToken.address,
+            amount,
+        };
+    },
+    async getSpender(chainId: number): Promise<string> {
+        try {
+            // 1Inch router contract address
+            const res = await retryHandling<OneInchSpenderRes>(() =>
+                httpClient.request<OneInchSpenderRes>(
+                    `${ONEINCH_SWAPS_ENDPOINT}${chainId}/approve/spender`
+                )
+            );
+
+            return res.address;
+        } catch (error) {
+            throw new Error('Unable to fetch exchange spender');
+        }
+    },
+    async getSwapQuote(
+        chainId: number,
+        params: SwapQuoteParams
+    ): Promise<SwapQuoteResponse> {
+        try {
+            const { fromTokenAddress, toTokenAddress, amount } =
+                this.parseQuoteParams(params);
+            const res = await retryHandling<OneInchSwapQuoteResponse>(() =>
+                httpClient.request<OneInchSwapQuoteResponse>(
+                    `${ONEINCH_SWAPS_ENDPOINT}${chainId}/quote`,
+                    {
+                        params: {
+                            fromTokenAddress,
+                            toTokenAddress,
+                            amount,
+                            fee: BASE_SWAP_FEE,
```

Comment:
why do we use `BASE_SWAP_FEE`  but then we use `BASE_SWAP_FEE * 10` in the return? should we use it in the request?

Reply:
BTW, using a constant with the value BASE_SWAP_FEE * 10/1000 could potentially reduce errors in the code

Reply:
`BASE_SWAP_FEE` is the percentage we sent to the API, representing the amount of `fromTokens` we'll get as a fee for the operation.

Currently, its value is **0.5**%, which means that if you swap 1000 USDC to X token, 5 USDC will be charged as fees.

The return will express those 5 USDC calculated in the following expression:
```
                blockWalletFee: BigNumber.from(res.fromTokenAmount)
                    .mul(BASE_SWAP_FEE * 10)
                    .div(1000)
```
1000 * (0,5 * 10) / 1000

This implementation was previously made, I just moved the code to a different file but I guess that the multiplication is used for better precision in the calculations with BigNumbers